src/tests/testCacheManager.cc

```diff
@@ -66,3 +75,86 @@ testCacheManager::testRegister()
     CPPUNIT_ASSERT_EQUAL(1,(int)sentry->flags);
 }
 
+void
+testCacheManager::testParseUrl()
+{
+    auto *mgr = static_cast<CacheManagerInternals *>(CacheManager::GetInstance());
+    CPPUNIT_ASSERT(mgr != nullptr);
+
+    std::vector<AnyP::ProtocolType> validSchemes = {
+        AnyP::PROTO_CACHE_OBJECT,
+        AnyP::PROTO_HTTP,
+        AnyP::PROTO_HTTPS,
+        AnyP::PROTO_FTP
+    };
+
+    AnyP::Uri mgrUrl;
+    mgrUrl.host("localhost");
+    mgrUrl.port(3128);
+
+    const std::vector<const char *> validPathActions = {
+        // "", // technically valid, but not supported
+        "/",
+        "/menu",
+        // "/squid-internal-mgr", // technically valid, but not supported
+        "/squid-internal-mgr/",
+        "/squid-internal-mgr/menu"
+    };
+
+    const std::vector<const char *> validParams = {
+        "",
+        "?",
+        "?&",
+        "?&&&&&&&&&&&&",
+        "?foo=bar",
+        "?0123456789=bar",
+        "?foo=bar&",
+        "?foo=bar&&&&",
+        "?&foo=bar",
+        "?&&&&foo=bar",
+        "?&foo=bar&",
+        "?&&&&foo=bar&&&&",
+        "?foo=?_weird?~`:[]stuff&bar=okay&&&&&&"
+    };
+
+    const std::vector<const char *> validFragments = {
+        "",
+        "#frag"
+    };
+
+    unsigned caseNumber = 0;
+    unsigned success = 0;
+    for (const auto &scheme : validSchemes) {
+        mgrUrl.setScheme(scheme);
+
+        for (const auto *action : validPathActions) {
+
+            // all schemes except cache_object require "/squid-internal-mgr" path prefix
+            if (scheme != AnyP::PROTO_CACHE_OBJECT && strlen(action) <= 19)
```

Comment:
We should not force ourselves to manually add the required magic prefix to nearly every action, decreasing the number of useful/interesting test cases and increasing noise. Please refactor to add the magic prefix to all actions (expecting a failure where it is not allowed) and remove the magic prefix from hard-coded validPathActions values.

----

If this code survives, we should not disassociate this code from its dependencies (and intent) so much! Please add something like a MgrUrMagic or a similar constant and use it here (and elsewhere, as appropriate).

```suggestion
            if (scheme != AnyP::PROTO_CACHE_OBJECT && strncmp(action, MgrUrlMagic, strlen(MgrUrlMagic)) != 0)
```

Reply:
Will do.

But please be aware that there are multiple magic prefix possible. cache_object:// only allowed "/" prefix. The current code uses "/squid-internal-mgr/". I also expect that we will need to register and move to a "/.well-known/" label since the Squid internal magic prefix are embedded in other peoples domains sometimes.

Reply:
Done in cd5118c4d